app/weather_forecast.py

```diff
@@ -0,0 +1,265 @@
+        location (str) - location name.
+    Returns:
+        (int) - minimum degrees in Celsius.
+        (int) - maximum degrees in Celsius.
+        (str) - weather conditions.
+        (str) - location / error description.
+    """
+    input_query_string = BASE_QUERY_STRING
+    input_query_string["startDateTime"] = input_date.isoformat()
+    input_query_string["endDateTime"] = (input_date + timedelta(days=1)).isoformat()
+    input_query_string["location"] = location
+    success_in_get_weather_data, api_json = get_data_from_api(HISTORY_URL, input_query_string)
+    if success_in_get_weather_data:
+        for item in api_json:
+            # print("historical:", api_json[item]['values'][0]['mint'], api_json[item]['values'][0]['maxt'])
+            min_temp = api_json[item]['values'][0]['mint']
+            max_temp = api_json[item]['values'][0]['maxt']
+            conditions = api_json[item]['values'][0]['conditions']
+        return min_temp, max_temp, conditions, api_json[item]['address']
+    else:
+        return None, None, None, api_json['message']
+
+
+def get_forecast_weather(input_date, location):
+    """ get the relevant weather forecast by calling the API.
+    Args:
+        input_date (date) - day part of date.
+        location (str) - location name.
+    Returns:
+        (int) - minimum degrees in Celsius.
+        (int) - maximum degrees in Celsius.
+        (str) - weather conditions.
+        (str) - location / error description.
+    """
+    input_query_string = BASE_QUERY_STRING
+    input_query_string["location"] = location
+    success_in_get_weather_data, api_json = get_data_from_api(FORECAST_URL, input_query_string)
+    if success_in_get_weather_data:
+        for item in api_json:
+            for i in range(len(api_json[item]['values'])):
+                if input_date == datetime.fromisoformat(api_json[item]['values'][i]['datetimeStr'][:-6]):
+                    min_temp = api_json[item]['values'][i]['mint']
+                    max_temp = api_json[item]['values'][i]['maxt']
+                    conditions = api_json[item]['values'][i]['conditions']
+        return min_temp, max_temp, conditions, api_json[item]['address']
+    else:
+        return None, None, None, api_json['message']
+
+
+def get_relevant_years_for_historical_average(day, month):
+    """ get a list for relevant years to call the get_historical_weather function
+        according to if date occurred this year or not.
+    Args:
+        day (int) - day part of date.
+        month (int) - month part of date.
+    Returns:
+        (list) - relevant years range.
+    """
+    if datetime.now() > datetime(year=datetime.now().year, month=month, day=day):
+        last_year = datetime.now().year
+    else:
+        last_year = datetime.now().year - 1
+    return list(range(last_year, last_year - HISTORICAL_AVERAGE_NUM_OF_YEARS, -1))
+
+
+def get_historical_average_weather(day, month, location):
```

Comment:
Unfortunately, this function is not scalable.

We can't send 50 API requests from the server for 3 reasons:

1. The API key will be depleted very fast.
2. If the request is [synchronous](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Synchronous_and_Asynchronous_Requests), it will cause the server to hang for a long time.
3. The load on the server gonna be huge for ~10k users.

There are two possible solutions: 

1. If there is an API service that supports this kind of historical querying, I suggest we use it.
2. If you can't find such an API service, we should (unfortunately) drop the historical average feature.

Reply:
Could not find a free API that provides that.
So I've changed the logic to find the last history data for the relevant day & month (best forecast)
29.02 is supported by trying to check if the date is valid for the relevant year & only if not we will try to get data for the 28'th.